### ISSUE TITLE
refine ignores solutions found by typeclass inference

### ISSUE THREAD

**Issue by jdchristensen** [coq/coq#18526]
#### Description of the problem

In the example below, `refine` has three subgoals.  Coq can solve two using typeclass inference, but discards the solutions for some reason.

```coq
Class Class1 (B : Type) := {}.

Class Class2 (B : Type) := {}.

Class Reflects (B : Type) := {}.

Definition O (B : Type) `{Reflects B} : Type.
Admitted.

Global Instance R_implies_C1 (B : Type) `{Reflects B} : Class1 (O B).
Admitted.

Definition lem (B : Type) {A : Type} `{Class1 B} `{Class2 B} : Class1 A.
Admitted.

Definition bad_behaviour {A B : Type} `{Reflects B} : Class1 A.
Proof.
  Typeclasses eauto := debug.
  simple refine (lem (O B)). (* log below *)
  1, 2: exact _. (* Coq finds these instances but then ignores them. *)
Abort.

(*
There are three goals, and we expect typeclass resolution to solve the first two.

Debug: Calling typeclass resolution with flags: depth = ∞,unique = false,fail = false
Debug: 1: looking for (Reflects B) without backtracking
Debug: 1.1: exact H on (Reflects B), 0 subgoal(s)
Debug: 2: looking for (Class1 (O B)) without backtracking
Debug: 2.1: simple apply R_implies_C1 on (Class1 (O B)), 0 subgoal(s)
Debug: 3: looking for (Class2 (O B)) without backtracking
Debug: 3: no match for (Class2 (O B)), 0 possibilities

That indeed happens, but when goal 3 fails, Coq discards the solutions to goals 1 and 2.
*)
```

This is a simplified version of something that came up in the Coq-HoTT library, see https://github.com/HoTT/Coq-HoTT/pull/1822

#### Coq Version

8.18.0, linux.  Also happens with 8.17.0, IIRC.  No other versions tested.

**Comment by Janno** [comment]
I think what's happening here is that `refine` gathers the subgoals that are typeclasses into a single typeclass query which either fails or succeeds. I suppose this behavior makes sense in general considering the fact that the subgoals could share evars and committing to a partial solution might make remaining subgoals unprovable.

If you need to work around this your best bet is probably to use `simple notypeclasses refine; [try apply _ ..]`.

As an aside: I personally would likely consider it a bug or at least code smell if lemmas whose hypotheses are typeclasses in `typeclass_instances` cannot have these hypotheses filled in automatically by typeclass search. Mostly because so many Coq tactics try to fill these arguments in manually and will probably fail completely when a TC subgoal cannot be solved.

**Comment by jdchristensen** [comment]
@Janno That explanation doesn't at all match my experience.  It's *very* common that a subset of the goals are solvable by typeclass search, and Coq does solve those, leaving the rest for the user.  And a slight modification of my example shows this:
```coq
Class Class1 (B : Type) := {}.

Class Class2 (B : Type) := {}.

Global Instance allclass1 (B : Type) : Class1 B.
Admitted.

Definition lem (A B : Type) `{Class1 B} `{Class2 B} : Class2 A.
Admitted.

Definition good_behaviour {A B : Type} : Class2 A.
Proof.
  simple refine (lem A B).
  (* [Class1 B] goal solved by typeclass search, but [Class2 B] goal left to do, as expected. *)
Abort.
```
The bad example above required some care to reproduce.  Lots of minor variations make it work as expected, so the failure is due to something a bit subtle, I think.

**Comment by jdchristensen** [comment]
I found an even simpler example showing the problem:
```coq
Class Class2 (B : Type) := {}.

Class Reflects (B : Type) := {}.

Definition O (B : Type) `{Reflects B} : Type.
Admitted.

Definition lem (B : Type) {A : Type} `{Class2 B} : Class2 A.
Admitted.

Definition bad_behaviour {A B : Type} `{Reflects B} : Class2 A.
Proof.
  Typeclasses eauto := debug.
  simple refine (lem (O B)). (* log below *)
  1: exact _. (* Coq finds this instance but then ignores it. *)
Abort.

(*
Debug: Calling typeclass resolution with flags: depth = ∞,unique = false,fail = false
Debug: 1: looking for (Reflects B) without backtracking
Debug: 1.1: exact H on (Reflects B), 0 subgoal(s)
Debug: 2: looking for (Class2 (O B)) without backtracking
Debug: 2: no match for (Class2 (O B)), 0 possibilities
*)
```
In the expression `lem (O B)`,  `O B` depends on `Reflects B`, which is a hypothesis, but it is not solved by typeclass search.

**Comment by Janno** [comment]
You are right. My explanation does not fit what is happening here. I don't know how much time I'll have today but I've started looking into the debug output with `Set Typeclasses Debug Verbosity 3` and it is already showing very unexpected differences between `good_behavior` the simplified `bad_behavior`. I think there might be a difference in how top-level underscores are treated versus how underscores in nested terms are treated by `refine`.

`good_behavior` seems to perform two TC searches back to back (although the debug output makes it look a lot like they are part of one invocation so it might be a matter of different options being passed that are not obvious in the debug output). Failure in the second subgoal does not (and cannot, I think) cause backtracking on the first subgoal.

`bad_behavior` starts out with basically the same relevant TC subgoals but here TC search decides to group them into a single query and failure in the second one causes backtracking in the first subgoal which does not have more hints to apply and thus fails.

Here's my current code. I tried making the examples as similar as possible:
```coq
Set Typeclasses Debug Verbosity 3.
Set Printing Existential Instances.
Set Printing All.
Module Expected.
  Class Class1 (B : Type) := {}.
  Class Class2 (B : Type) := {}.
  Global Instance allclass1 (B : Type) : Class1 B. Admitted.
  Axiom lem : forall (A B : Type) `(Class1 B) `(Class2 B), False.
  Definition good_behaviour {A B : Type} : False.
  Proof.
    simple refine (@lem A B _ _).
    (* [Class1 B] goal solved by typeclass search, but [Class2 B] goal left to do, as expected. *)
  Abort.
End Expected.
Module Unexpected.
  Class Class1 (B : Type) := {}.
  Class Class2 (B : Type) := {}.
  Global Instance allclass1 B : Class2 B. Admitted.
  Axiom O : forall (B : Type) `(Class2 B), Type.
  Axiom lem : forall (B : Type) (A : Type) `(Class1 B), False.
  Definition bad_behaviour {A B : Type} : False.
  Proof.
    simple refine (@lem (O B _) _ _).
    (* [Class1 B] succeeds but [Class2 B] fails and makes [Class1 B] backtrack and then fail. *)
  Abort.
End Unexpected.
```

**Comment by SkySkimmer** [comment]
>[Class1 B] succeeds but [Class2 B] fails and makes [Class1 B] backtrack and then fail.

There's no `Class1 B` involved, it's `Class1 (O B _)`. So it's "[Class2 B] succeeds but [Class1 (O B (the result of Class2 B))] fails".

At the start of resolution we see that the two goals are related (the Class1 goal depends on Class2) so when one fails the whole group fails (we may have made a choice that caused the failure).

>I think what's happening here is that refine gathers the subgoals that are typeclasses into a single typeclass query which either fails or succeeds. I suppose this behavior makes sense in general considering the fact that the subgoals could share evars and committing to a partial solution might make remaining subgoals unprovable.

It's not about refine specifically, it's typeclass search on multiple goals. It groups them in independent components.

**Comment by jdchristensen** [comment]
@SkySkimmer Thanks for explaining the logic.  Is that behaviour configurable?

What really confused me is that I thought that in the original place where this arose, I saw behaviour that contradicts what you say.  A typeclass instance is found, and a typeclass goal that depends on it is left.  But, looking more closely, I think what is happening is that the typeclass instance that is found is found via unification, not typeclass search, so while it superficially looks like a counterexample, it is not.  (It's hard to see for sure, because there are coercions at work as well.)

In any case, the main point for me is that I learned more about how typeclass search works.  I appreciate the explanation.  Feel free to close.

**Comment by Alizter** [comment]
I think rather than closing this issue, it could be turned into a documentation opportunity about how typeclasses are treated vs unification in tactics.

**Comment by SkySkimmer** [comment]
It's not configurable.

**Comment by SkySkimmer** [comment]
Not configurable AFAIK.

**Comment by SkySkimmer** [comment]
There's a `best_effort` variant but it seems to be about hint mode mismatches not this.
